### DISCUSSION TITLE
Reverse migration when disabling a feature

### DISCUSSION THREAD

**Discussion by dealdiane** [OrchardCMS/OrchardCore#10110] [Q&A] [answered]
Is there a convention that I can use to run an "undo" migration when a feature is disabled? There is an "uninstall" lookup in the out-of-the-box `IDataMigrationManager` implementation (`DataMigrationManager`) but no undo/revert/reverse/downgrade lookup if I'm not wrong. Also, "uninstall" is only called when uninstalling features anyway.

The way I have it setup so far is by implementing the `IFeatureEventHandler` interface and manually executing my migrations when my feature is disabled. Untidy but kind of works. I'm hoping that there's a better way to do it. Any advice?

**Comment by Piedone** [comment, marked as answer]
Currently, your best bet is what you're already doing.

We could actually have uninstallation surfaced properly, see this issue: https://github.com/OrchardCMS/OrchardCore/issues/2562. I'd argue that such "reverse" operations are actually more appropriate for an uninstall operation (i.e. when you want to get rid of the feature) than a disable (after which you may soon re-enable it).

**Reply by dealdiane** [reply]
My problem is with a content part that gets automatically added when enabling a feature. I want this part removed (or, if there's a way, at least disabled) when the feature is disabled. Because the part needs the drivers and a few other things in the module, the part is essentially useless without them anyway.

Also, a different scenario but this is actually how I came across this is when I accidentally removed the part from the admin. It was accidental so I wanted it added back but there's no way to rollback and re-run the "create" migration unless explicitly calling "uninstall". If there was an uninstall (should be addressed by #2562) or rolling back is supported when a feature is disabled, it would've been a lot easier.

**Reply by Piedone** [reply]
When talking about parts, specifically, you can try to add them to items of given content types instead of making them attachable and attaching them by hand or via migrations statically.

I'm a bit unsure on the details but I think if you do `contentItem.Weld<MyPart>()` in a handler in `ActivatingAsync()` or `LoadingAsync`. You can do this conditionally based on what's in the content item object (like checking its content type). This will make your content part work pretty much like if you've attached it to the content type via migrations. Note that you won't have content part settings available for configuration under the content type editor though. Then, since your handle will run only when your feature is enabled, your part will be attached only when your feature is active.

This way you can't remove the part from the type accidentally either.

**Reply by dealdiane** [reply]
That's an interesting way of doing it and it makes sense to "Weld" it instead, at least from our current business requirement's perspective. 

I will look at maybe redesigning it in the future but for now, we need the part settings and it works - as long as the part doesn't get deleted. I'm sure I can find a way to manage the settings somewhere else. Like in the SiteSettings maybe. That may not work for everyone, but again, for our requirements, SiteSettings may be sufficient.

Thanks
